src/provider/form-factory-provider.tsx

```diff
@@ -0,0 +1,138 @@
+import React, { createContext, useContext, useEffect, useRef } from 'react';
+import { type FormField, type FormSchema, type SessionMode } from '../types';
+import { EncounterFormProcessor } from '../processors/encounter/encounter-form-processor';
+import { type LayoutType, useLayoutType, type OpenmrsResource } from '@openmrs/esm-framework';
+import { type FormProcessorConstructor } from '../processors/form-processor';
+import { type FormContextProps } from './form-provider';
+import { validateForm } from './form-factory-helper';
+
+interface FormFactoryProviderContextProps {
+  patient: fhir.Patient;
+  sessionMode: SessionMode;
+  sessionDate: Date;
+  formJson: FormSchema;
+  formProcessors: Record<string, FormProcessorConstructor>;
+  layoutType: LayoutType;
+  workspaceLayout: 'minimized' | 'maximized';
+  visit: OpenmrsResource;
+  location: OpenmrsResource;
+  provider: OpenmrsResource;
+  registerForm: (formId: string, context: FormContextProps, isSubForm: boolean) => void;
+  getSubForms: () => FormContextProps[];
+  getRootForm: () => FormContextProps;
+  setCurrentPage: (page: string) => void;
+  handleConfirmQuestionDeletion?: (question: Readonly<FormField>) => Promise<void>;
+}
+
+interface FormFactoryProviderProps {
+  patient: fhir.Patient;
+  sessionMode: SessionMode;
+  sessionDate: Date;
+  formJson: FormSchema;
+  workspaceLayout: 'minimized' | 'maximized';
+  location: OpenmrsResource;
+  provider: OpenmrsResource;
+  visit: OpenmrsResource;
+  children: React.ReactNode;
+  formSubmissionProps: {
+    isSubmitting: boolean;
+    setIsSubmitting: (isSubmitting: boolean) => void;
+    onSubmit: (data: any) => void;
+    onError: (error: any) => void;
+    handleClose: () => void;
+  };
+  setCurrentPage: (page: string) => void;
+  handleConfirmQuestionDeletion?: (question: Readonly<FormField>) => Promise<void>;
+}
+
+const FormFactoryProviderContext = createContext<FormFactoryProviderContextProps | undefined>(undefined);
+
+export const FormFactoryProvider: React.FC<FormFactoryProviderProps> = ({
+  patient,
+  sessionMode,
+  sessionDate,
+  formJson,
+  workspaceLayout,
+  location,
+  provider,
+  visit,
+  children,
+  formSubmissionProps,
+  setCurrentPage,
+  handleConfirmQuestionDeletion,
+}) => {
+  const rootForm = useRef<FormContextProps>();
+  const subForms = useRef<Record<string, FormContextProps>>({});
+  const layoutType = useLayoutType();
+  const { isSubmitting, setIsSubmitting, onSubmit, onError, handleClose } = formSubmissionProps;
+  const registerForm = (formId: string, context: FormContextProps, isSubForm: boolean) => {
```

Comment:
This should probably be wrapped in `useCallback()`

Reply:
I also wonder about the `isSubForm` parameter. Shouldn't the assumption be something like the first form registered is the root form and every other form is a subform, so something like:

```ts
if (!rootForm.current) {
  rootForm.current = context;
} else {
  subForm.current[formId] = context;
}
```

Reply:
Nice one!